Add information about different types of ACS (#1971)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-rhui-alternate-content-sources.adoc

[id="Configuring_RHUI_Alternate_Content_Sources_{context}"]
= Configuring RHUI Alternate Content Sources

.Procedure
. In the {ProjectWebUI}, navigate to *Content* > *Alternate Content Sources*.
. Click *Add Source* and set the *Source type* as *RHUI*.
. Generate RHUI certificates using the command provided in the {ProjectWebUI}.
Ensure that you pass the repo labels of the desired repositories.
. In the *Name* field, enter a name for the alternate content source.
. Optional: In the the *Description* field, provide a description for the ACS.

> . Select {SmartProxies} to which the alternate content source is to be synced.

Reviewer: Also: `. Optional: Select 'Use HTTP proxies' if you want the ACS to use the {SmartProxy}'s HTTP proxy.`

[id="Configuring_RHUI_Alternate_Content_Sources_{context}"]

> = Configuring RHUI Alternate Content Sources

Reviewer 2: What is RHUI? For now, all ACS procedures can be part of all build targets IMHO. In downstream, we don't run orcharhino on Foreman 3.5+ yet.

Reviewer 3: Red Hat Update Infrastructure. It's the service deployed by Red Hat inside clouds such as AWS, Azure and Google to have the Red Hat CDN available locally, without additional bandwidth costs.

.Procedure
. In the {ProjectWebUI}, navigate to *Content* > *Alternate Content Sources*.
. Click *Add Source* and set the *Source type* as *RHUI*.
. Generate RHUI certificates using the command provided in the {ProjectWebUI}.
Ensure that you pass the repo labels of the desired repositories.
. In the *Name* field, enter a name for the alternate content source.
. Optional: In the the *Description* field, provide a description for the ACS.
. Select {SmartProxies} to which the alternate content source is to be synced.
. Optional: Select `Use HTTP proxies` if you want the ACS to use the {SmartProxy}'s HTTP proxy.
. Enter the Base URL of the Red Hat Update Infrastructure server.
. Enter a comma-separated list of Subpaths.
. Provide the *Content Credentials*, if these are needed.
. If SSL verification is required, enable *Verify SSL* and select the SSL CA certificate.
. Review details and click *Add*.

> . Navigate to *Content* > *Alternate Content Sources* > click the vertical ellipsis next to the newly created alternate content source > Select *Refresh*.

Reviewer 4: Do we need to add the 'CLI Procedure' also here?

Author: @vsedmik @ianballou  done!

[id="Configuring_RHUI_Alternate_Content_Sources_{context}"]
= Configuring RHUI Alternate Content Sources

.Prerequisites
* Generate the client entitlement certificates for the required repos on the RHUA node as described in https://access.redhat.com/documentation/en-us/red_hat_update_infrastructure/4/html/configuring_and_managing_red_hat_update_infrastructure/assembly_cmg-creating-client-ent-cert-config-rpm_configuring-and-managing-red-hat-update-infrastructure#proc_cmg-creating-client-entitlement-certificate_assembly_cmg-creating-client-ent-cert-config-rpm[Creating a client entitlement certificate with the Red Hat Update Infrastructure Management Tool] in the _Configuring and Managing Red Hat Update Infrastructure Guide_.

> Import the client entitlement certificates to the {Project}.

Reviewer 2: ```suggestion
* Import the client entitlement certificates into your {Project}.
```

For more information, see {ContentManagementDocURL}Importing_Custom_SSL_Certificates_content-management[Importing Custom SSL Certificates] in _{ContentManagementDocTitle}_.
* Obtain a list of the subpaths for the required repositories.
Execute the following command on the RHUA server:
+
[options="nowrap" subs="+quotes,attributes"]
----
# rhui-manager repo info --repo_id _My_Repo_ID_
----

.Procedure
. In the {ProjectWebUI}, navigate to *Content* > *Alternate Content Sources*.
. Click *Add Source* and set the *Source type* as *RHUI*.
. Generate RHUI certificates using the command provided in the {ProjectWebUI}.
Ensure that you pass the repo labels of the desired repositories.
. In the *Name* field, enter a name for the alternate content source.

> . Optional: In the the *Description* field, provide a description for the ACS.

Reviewer 5: ```suggestion
. Optional: In the *Description* field, provide a description for the ACS.
```

[id="Configuring_RHUI_Alternate_Content_Sources_{context}"]
= Configuring RHUI Alternate Content Sources

.Prerequisites
* Generate the client entitlement certificates for the required repos on the RHUA node as described in https://access.redhat.com/documentation/en-us/red_hat_update_infrastructure/4/html/configuring_and_managing_red_hat_update_infrastructure/assembly_cmg-creating-client-ent-cert-config-rpm_configuring-and-managing-red-hat-update-infrastructure#proc_cmg-creating-client-entitlement-certificate_assembly_cmg-creating-client-ent-cert-config-rpm[Creating a client entitlement certificate with the Red Hat Update Infrastructure Management Tool] in the _Configuring and Managing Red Hat Update Infrastructure Guide_.
* Import the client entitlement certificates into your {Project}.
For more information, see {ContentManagementDocURL}Importing_Custom_SSL_Certificates_content-management[Importing Custom SSL Certificates] in _{ContentManagementDocTitle}_.
* Obtain a list of the subpaths for the required repositories.
Execute the following command on the RHUA server:
+
[options="nowrap" subs="+quotes,attributes"]
----
# rhui-manager repo info --repo_id _My_Repo_ID_
----

> * Note that the alternate content source paths consist of a base URL appended with the subpaths that you provide. For example, if your base URL is `https://server.example.com` and your subpaths are `rhel7/` and `rhel8/`, then both `https://server.example.com/rhel7/` and `https://server.example.com/rhel8/` will be searched. 

Reviewer 2: ```suggestion
* Note that the alternate content source paths consist of a base URL appended with the subpaths that you provide.
For example, if your base URL is `https://server.example.com` and your subpaths are `rhel7/` and `rhel8/`, then both `https://server.example.com/rhel7/` and `https://server.example.com/rhel8/` will be searched. 
```

Ensure that you pass the repo labels of the desired repositories.
. In the *Name* field, enter a name for the alternate content source.
. Optional: In the *Description* field, provide a description for the ACS.
. Select {SmartProxies} to which the alternate content source is to be synced.
. Optional: Select `Use HTTP proxies` if you want the ACS to use the {SmartProxy}'s HTTP proxy.
. Enter the Base URL of the Red Hat Update Infrastructure CDS node.
. Enter a comma-separated list of Subpaths.
. Provide the *Content Credentials*, if these are needed.
. If SSL verification is required, enable *Verify SSL* and select the SSL CA certificate.
. Review details and click *Add*.
. Navigate to *Content* > *Alternate Content Sources* > click the vertical ellipsis next to the newly created alternate content source > Select *Refresh*.

[id="cli-configuring-rhui-alternate-content-sources_{context}"]
.CLI Procedure
. On {ProjectServer}, enter the following command:
+
[options="nowrap" subs="+quotes,attributes"]
----
# hammer alternate-content-source create \
--name "_My_ACS_Name_" \
--alternate-content-source-type rhui \

> --base-url "_https://local-repo.example.com:port_" \

Reviewer 2: What do you mean by "local-repo"? Is it just any server/mirror?

Author: It could be any server that hosts content from a repo, on the local network, or a repo mirror.

Reviewer 2: Maybe reword to "mirror.example.com"?

Author: @maximiliankolb now that I think of it, I feel that the baseURL here should be the RHUI CDS node. @ianballou @sjha4 can you guys confirm? We can relabel the placeholder URL likewise.

. Optional: In the *Description* field, provide a description for the ACS.
. Select {SmartProxies} to which the alternate content source is to be synced.
. Optional: Select `Use HTTP proxies` if you want the ACS to use the {SmartProxy}'s HTTP proxy.
. Enter the Base URL of the Red Hat Update Infrastructure CDS node.
. Enter a comma-separated list of Subpaths.
. Provide the *Content Credentials*, if these are needed.
. If SSL verification is required, enable *Verify SSL* and select the SSL CA certificate.
. Review details and click *Add*.
. Navigate to *Content* > *Alternate Content Sources* > click the vertical ellipsis next to the newly created alternate content source > Select *Refresh*.

[id="cli-configuring-rhui-alternate-content-sources_{context}"]
.CLI Procedure
. On {ProjectServer}, enter the following command:
+
[options="nowrap" subs="+quotes,attributes"]
----
# hammer alternate-content-source create \
--name "_My_ACS_Name_" \
--alternate-content-source-type rhui \
--base-url "_https://rhui-cds-node/pulp/content_" \
--subpaths _path/to/repo/1/, path/to/repo/2/_ \
--ssl-client-cert-id _SSL_Client_Cert_ID_ \
--ssl-client-key-id _SSL_Client_Key_ID_ \

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer 3: IMHO the project/product should always verify SSL by default. Is there a particular reason you don't have this default?

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer: We did have it as a default before.  It got turned into a required argument in the API when we fixed a bug around verify_ssl being passed silently into the Simplified ACS type, which shouldn't accept it as parameter.

Long story short, it's a bug that we're now requiring verify_ssl. However it'll be that way in the next Katello release so I figured we should document it as such for now.

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer: @qcjames53 I'm trying to remember, do we have a Redmine to reinstate the ACS verify-ssl True default?

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer 3: That's fair. I think it's a good practice to link the the bug (Redmine, GH issue) in a code comment. That makes it easier to recognize it's a workaround and to see if it can already be dropped.

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer 6: @ianballou The talk we had with Jeremy was inconclusive. I believe we came to the conclusion that we need to leave verify-ssl as a required parameter for now unless the custom/rhui and simplified ACS's are split into separate classes.

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer 6: There's no redmine at the moment.

> --smart-proxy-ids _{SmartProxy}_ID_

Reviewer: For completeness, here is the redmine: https://projects.theforeman.org/issues/36187

[id="Configuring_RHUI_Alternate_Content_Sources_{context}"]
= Configuring RHUI Alternate Content Sources

.Prerequisites

> * Generate the client entitlement certificates for the required repos on the RHUA node as described in https://access.redhat.com/documentation/en-us/red_hat_update_infrastructure/4/html/configuring_and_managing_red_hat_update_infrastructure/assembly_cmg-creating-client-ent-cert-config-rpm_configuring-and-managing-red-hat-update-infrastructure#proc_cmg-creating-client-entitlement-certificate_assembly_cmg-creating-client-ent-cert-config-rpm[Creating a client entitlement certificate with the Red Hat Update Infrastructure Management Tool] in the _Configuring and Managing Red Hat Update Infrastructure Guide_.

Reviewer 2: ```suggestion
* Generate the client entitlement certificates for the required repos on the RHUA node as described in https://access.redhat.com/documentation/en-us/red_hat_update_infrastructure/4/html/configuring_and_managing_red_hat_update_infrastructure/assembly_cmg-creating-client-ent-cert-config-rpm_configuring-and-managing-red-hat-update-infrastructure#proc_cmg-creating-client-entitlement-certificate_assembly_cmg-creating-client-ent-cert-config-rpm[Creating a client entitlement certificate with the Red Hat Update Infrastructure Management Tool] in _Configuring and Managing Red Hat Update Infrastructure_.
```

What does RHUA mean?

* Import the client entitlement certificates into your {Project}.
For more information, see {ContentManagementDocURL}Importing_Custom_SSL_Certificates_content-management[Importing Custom SSL Certificates] in _{ContentManagementDocTitle}_.
* Obtain a list of the subpaths for the required repositories.

> Execute the following command on the RHUA server:

Reviewer 2: ```suggestion
Execute the following command on your RHUA server:
```

+
[options="nowrap" subs="+quotes,attributes"]
----
# rhui-manager repo info --repo_id _My_Repo_ID_
----
* Note that the alternate content source paths consist of a base URL appended with the subpaths that you provide.

> For example, if your base URL is `https://server.example.com` and your subpaths are `rhel7/` and `rhel8/`, then both `https://server.example.com/rhel7/` and `https://server.example.com/rhel8/` will be searched. 

Reviewer 2: same as above: use attribute if it makes sense.

Reviewer: I commented on the above one too

. In the *Name* field, enter a name for the alternate content source.
. Optional: In the *Description* field, provide a description for the ACS.
. Select {SmartProxies} to which the alternate content source is to be synced.
. Optional: Select `Use HTTP proxies` if you want the ACS to use the {SmartProxy}'s HTTP proxy.
. Enter the Base URL of the Red Hat Update Infrastructure CDS node.
. Enter a comma-separated list of Subpaths.
. Provide the *Content Credentials*, if these are needed.
. If SSL verification is required, enable *Verify SSL* and select the SSL CA certificate.
. Review details and click *Add*.
. Navigate to *Content* > *Alternate Content Sources* > click the vertical ellipsis next to the newly created alternate content source > Select *Refresh*.

[id="cli-configuring-rhui-alternate-content-sources_{context}"]
.CLI Procedure
. On {ProjectServer}, enter the following command:
+
[options="nowrap" subs="+quotes,attributes"]
----
# hammer alternate-content-source create \
--name "_My_ACS_Name_" \
--alternate-content-source-type rhui \
--base-url "_https://rhui-cds-node/pulp/content_" \
--subpaths _path/to/repo/1/,path/to/repo/2/_ \

> --ssl-client-cert-id _SSL_Client_Cert_ID_ \
> --ssl-client-key-id _SSL_Client_Key_ID_ \
> --smart-proxy-ids _{SmartProxy}_ID_ \

Reviewer 2: ```suggestion
--ssl-client-cert-id _My_SSL_Client_Certificate_ID_ \
--ssl-client-key-id _My_SSL_Client_Key_ID_ \
--smart-proxy-ids _My_{SmartProxy}_ID_ \
```
